File: Player/TableViewPlayer/objc/TableViewPlayer/VideoTableViewCell.m
//
//  VideoTableViewCell.m
//  TableViewPlayer
//
//  Copyright © 2024 Brightcove, Inc. All rights reserved.
//

#import "Notifications.h"
#import "PlaybackConfiguration.h"

#import "VideoTableViewCell.h"


@interface VideoTableViewCell ()

@property (nonatomic, weak) IBOutlet UIView *videoContainerView;
@property (nonatomic, weak) IBOutlet UILabel *videoLabel;
@property (nonatomic, weak) IBOutlet UIButton *muteButton;

@property (nonatomic, weak) PlaybackConfiguration *playbackConfiguration;
@property (nonatomic, strong) BCOVPUIPlayerView *playerView;

@end


@implementation VideoTableViewCell

#pragma mark - Lifecycle

- (void)prepareForReuse
{
    [super prepareForReuse];
    [self.playbackConfiguration.playbackController pause];
    self.playbackConfiguration.playbackSession.player.muted = YES;
    [self updateMuteButton];
}

- (void)awakeFromNib
{
    [super awakeFromNib];

    self.playerView = ({
        BCOVPUIPlayerViewOptions *options = [BCOVPUIPlayerViewOptions new];
        BCOVPUIBasicControlView *controlsView = BCOVPUIBasicControlView.basicControlViewWithVODLayout;

        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"self.tag == %@", @(BCOVPUIViewTagButtonScreenMode)];
        BCOVPUILayoutView *screenModeButton = [controlsView.layout.allLayoutItems.allObjects filteredArrayUsingPredicate:predicate].firstObject;
        screenModeButton.removed = YES;

        BCOVPUIPlayerView *playerView = [[BCOVPUIPlayerView alloc]
                                         initWithPlaybackController:nil
                                         options:options
                                         controlsView:controlsView];

        playerView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        playerView.frame = self.videoContainerView.bounds;
        [self.videoContainerView addSubview:playerView];

        playerView;
    });

    // Handle when another video is unmuted
    [NSNotificationCenter.defaultCenter addObserver:self
                                           selector:@selector(unmuteNotificationReceived:)
                                               name:VideoDidUnmuteNotification
                                             object:nil];

    // Handle when the table view stops scrolling
    // We want to play videos in when this happens
    [NSNotificationCenter.defaultCenter addObserver:self
                                           selector:@selector(scrollingStoppedNotificationReceived:)
                                               name:ScrollingStoppedNotification object:nil];

    // Handle when the table view starts scrolling
    // We want to pause videos when this happens
    [NSNotificationCenter.defaultCenter addObserver:self
                                           selector:@selector(scrollingStartedNotificationReceived:)
                                               name:ScrollingStartedNotification
                                             object:nil];
}


- (void)setUpWithVideo:(BCOVVideo *)video
 playbackConfiguration:(PlaybackConfiguration *)playbackConfiguration
{
    self.playbackConfiguration = playbackConfiguration;
    self.playerView.playbackController = playbackConfiguration.playbackController;
    self.videoLabel.text = video.properties[kBCOVVideoPropertyKeyName];
}

- (void)unmuteNotificationReceived:(NSNotification *)notification
{
    // Mute all videos except the one for this cell
    UITableViewCell *cell = notification.object;
    if (![cell isEqual:self])
    {
        self.playbackConfiguration.playbackSession.player.muted = YES;
        [self updateMuteButton];
    }
}

- (void)updateMuteButton
{
    AVPlayer *player = self.playbackConfiguration.playbackSession.player;

    NSString *title = player.isMuted ? @"Unmute" : @"Mute";
    [self.muteButton setTitle:title
                     forState:UIControlStateNormal];
}

- (void)scrollingStoppedNotificationReceived:(NSNotification *)notification
{
    [self.playbackConfiguration.playbackController play];
}

- (void)scrollingStartedNotificationReceived:(NSNotification *)notification
{
    [self.playbackConfiguration.playbackController pause];
}

- (IBAction)toggleVideoMute:(id)sender
{
    AVPlayer *player = self.playbackConfiguration.playbackSession.player;

    player.muted = !player.isMuted;

    [self updateMuteButton];

    if (!player.isMuted)
    {
        [NSNotificationCenter.defaultCenter postNotificationName:VideoDidUnmuteNotification
                                                          object:self];
    }
}

@end

File: Player/TableViewPlayer/swift/TableViewPlayer/VideoTableViewCell.swift
//
//  VideoTableViewCell.swift
//  TableViewPlayer
//
//  Copyright © 2024 Brightcove, Inc. All rights reserved.
//

import Foundation
import UIKit
import BrightcovePlayerSDK


let UnmuteNotification = Notification.Name("VideoDidUnmute")


final class VideoTableViewCell: UITableViewCell {
    
    @IBOutlet fileprivate weak var videoContainerView: UIView!
    @IBOutlet fileprivate weak var videoLabel: UILabel!
    @IBOutlet fileprivate weak var muteButton: UIButton! {
        didSet {
            muteButton.tintColor = .black
        }
    }
    
    fileprivate lazy var playerView: BCOVPUIPlayerView? = {
        let options = BCOVPUIPlayerViewOptions()
        
        let controlsView = BCOVPUIBasicControlView.withVODLayout()
        if let allLayoutItems = controlsView?.layout.allLayoutItems,
           let screenModeButton = allLayoutItems.first(where: { ($0 as? BCOVPUILayoutView)?.tag == BCOVPUIViewTag.buttonScreenMode.rawValue }) as? BCOVPUILayoutView {
            screenModeButton.isRemoved = true
        }
        
        guard let playerView = BCOVPUIPlayerView(playbackController: nil,
                                                 options: options,
                                                 controlsView: controlsView) else {
            return nil
        }
        
        playerView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        playerView.frame = videoContainerView.bounds
        videoContainerView.addSubview(playerView)
        
        return playerView
    }()
    
    fileprivate weak var playbackConfiguration: PlaybackConfiguration?
    
    override func prepareForReuse() {
        super.prepareForReuse()
        
        playbackConfiguration?.playbackController?.pause()
        if let player = playbackConfiguration?.playbackSession?.player {
            player.isMuted = true
            updateMuteButton()
        }
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        // Handle when another video is unmuted
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(unmuteNotificationReceived(_:)),
                                               name: UnmuteNotification,
                                               object: nil)
        
        // Handle when the table view stops scrolling
        // We want to play videos in when this happens
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(scrollingStoppedNotificationReceived(_:)),
                                               name: ScrollingStoppedNotification,
                                               object: nil)
        
        // Handle when the table view starts scrolling
        // We want to pause videos when this happens
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(scrollingStartedNotificationReceived(_:)),
                                               name: ScrollingStartedNotification,
                                               object: nil)
    }
    
    func setUpWith(video: BCOVVideo,
                   playbackConfiguration: PlaybackConfiguration) {
        self.playbackConfiguration = playbackConfiguration
        
        if let playerView,
           let playbackController = playbackConfiguration.playbackController {
            playerView.playbackController = playbackController
        }
        
        if let videoTitle = video.properties[BCOVVideo.PropertyKeyName] as? String {
            videoLabel.text = videoTitle
        }
    }
    
    @objc
    fileprivate func unmuteNotificationReceived(_ notification: Notification) {
        guard let player = playbackConfiguration?.playbackSession?.player,
              let cell = notification.object as? VideoTableViewCell else {
            return
        }
        
        // Mute all videos except the one for this cell
        if cell != self {
            player.isMuted = true
            updateMuteButton()
        }
    }
    
    fileprivate func updateMuteButton() {
        guard let player = playbackConfiguration?.playbackSession?.player else {
            return
        }
        
        let title = player.isMuted ? "Unmute" : "Mute"
        muteButton.setTitle(title, for: .normal)
    }
    
    @objc
    fileprivate func scrollingStoppedNotificationReceived(_ notification: Notification) {
        playbackConfiguration?.playbackController?.play()
    }
    
    @objc
    fileprivate func scrollingStartedNotificationReceived(_ notification: Notification) {
        playbackConfiguration?.playbackController?.pause()
    }
    
    @IBAction
    fileprivate func toggleVideoMute() {
        guard let player = playbackConfiguration?.playbackSession?.player else {
            return
        }
        
        player.isMuted = !player.isMuted
        
        updateMuteButton()
        
        if !player.isMuted {
            NotificationCenter.default.post(name: UnmuteNotification,
                                            object: self)
        }
    }
}
